Add some clarifying language to the semantics of metric instrument naming (#552)
open-telemetry/opentelemetry-specification: specification/metrics/api-user.md

  This name will be used as the `namespace` for any metrics created using the returned `Meter`.
- `version` (optional): Specifies the version of the instrumentation library (e.g. `semver:1.0.0`).

### Metric Instrument names

Metric instruments have names, which are how we refer to them in
external systems.  Metric instrument names conform to the following syntax:

1. They are non-empty strings
2. They are case-insensitive
3. The first character must be non-numeric, non-space, non-punctuation
4. Subsequent characters must be belong to the alphanumeric characters, '_', '.', and '-'.

> Metric instrument names belong to a namespace, which is the `name` of the associated `Meter`,

Reviewer: We discussed yesterday during metrics sig about this, I think the `meter name` is not a `namespace`, more like a `source`. Also I heard on gitter that you want to followup on this, what do you want to do with this PR then?

Author: I'd like to leave this PR open; one way or another, clarification is needed.

Reviewer 2: @bogdandrutu We adapted the spec to explicitly use the meter name as namespace a while ago after it was proposed in a SIG spec meeting since this puts the metric names into a namespace per library to avoid name clashes between libraries unaware of each other. See https://github.com/open-telemetry/opentelemetry-specification/issues/391#issuecomment-566299849, #408 and https://github.com/open-telemetry/oteps/pull/76#issuecomment-574846369. What has changed since then?

Reviewer: @arminru do you still have concerns?

Reviewer 2: I thought you wanted to revise this and depart from using the meter name as a namespace / collision domain entirely but maybe I was missing context or misunderstood.
I'm fine with the current wording this PR introduces and therefore had already approved it.

allowing the same metric name to be used in multiple libraries of code,
unambiguously, within the same application.

Metric instrument names SHOULD be semantically meaningful, even when viewed
outside of the context of the originating Meter name. For example, when instrumenting
an http server library, "latency" is not an appropriate instrument name, as it is too generic.
Instead, favor a name like "http_request_latency", as it informs the viewer of the

> semantic meaning of the latency.

Reviewer: Just add an extra comment to say that this is an example and the list of standard names will be defined separately. I try to avoid making this example the canonical name for `http_request_latency` (not because it is a bad name, but because every document may include some examples and we want to have a semantic convention separate doc to keep all the names in one place).

Author: I think until we have semantic conventions, we need some guidance, via examples, no?

Reviewer 3: from the spec sig mtg today, sounds like the example just needs to be touched up so the example is not interpreted as authoritative cc @bogdandrutu @jkwatson

Author: I updated the language in here. PTAL

Metric instruments are defined using a `Meter` instance, using a variety
of `New` methods specific to the kind of metric and type of input (integer
or floating point).  The Meter will return an error when a metric name is

> already registered with a different kind for the same name. Metric systems

Reviewer 4: ```suggestion
already registered with a different kind for the same name. Metric exporting systems
```

> are expected to automatically prefix exported metrics by the namespace, if

Reviewer 4: We probably want to rephrase this to not prescribe a solution, but instead assign responsibility.

```suggestion
are responsible for disambiguation in the context of the namespace, if
```
